book/sources/exex/remote/src/exex.rs

```diff
@@ -0,0 +1,88 @@
+use futures_util::StreamExt;
+use remote_exex::proto::{
+    self,
+    remote_ex_ex_server::{RemoteExEx, RemoteExExServer},
+};
+use reth_exex::{ExExContext, ExExEvent, ExExNotification};
+use reth_node_api::FullNodeComponents;
+use reth_node_ethereum::EthereumNode;
+use reth_tracing::tracing::info;
+use std::sync::Arc;
+use tokio::sync::{broadcast, mpsc};
+use tokio_stream::wrappers::ReceiverStream;
+use tonic::{transport::Server, Request, Response, Status};
+
+struct ExExService {
+    notifications: Arc<broadcast::Sender<ExExNotification>>,
+}
+
+#[tonic::async_trait]
+impl RemoteExEx for ExExService {
+    type SubscribeStream = ReceiverStream<Result<proto::ExExNotification, Status>>;
+
+    async fn subscribe(
+        &self,
+        _request: Request<proto::SubscribeRequest>,
+    ) -> Result<Response<Self::SubscribeStream>, Status> {
+        let (tx, rx) = mpsc::channel(1);
+
+        let mut notifications = self.notifications.subscribe();
+        tokio::spawn(async move {
+            while let Ok(notification) = notifications.recv().await {
+                let proto_notification = proto::ExExNotification {
+                    data: bincode::serialize(&notification).expect("failed to serialize"),
+                };
+                tx.send(Ok(proto_notification))
+                    .await
+                    .expect("failed to send notification to client");
+
+                info!("Notification sent to the gRPC client");
+            }
+        });
+
+        Ok(Response::new(ReceiverStream::new(rx)))
+    }
+}
+
+async fn remote_exex<Node: FullNodeComponents>(
+    mut ctx: ExExContext<Node>,
+    notifications: Arc<broadcast::Sender<ExExNotification>>,
+) -> eyre::Result<()> {
+    while let Some(result) = ctx.notifications.next().await {
+        let notification = result?;
+        if let Some(committed_chain) = notification.committed_chain() {
+            ctx.events.send(ExExEvent::FinishedHeight(committed_chain.tip().num_hash()))?;
+        }
+
+        info!("Notification sent to the gRPC server");
+        let _ = notifications.send(notification);
+    }
+
+    Ok(())
+}
+
+// ANCHOR: snippet
```

Comment:
do we need this?

Reply:
I wanted to show only the relevant sections of the files at the book references, and allow the reader to expand the complete file only if they want to. Another way to do this is to use line numbers of the snippets instead of using these `// ANCHOR: snippet` comments, but then we'd have to change the line numbers if the examples are changed. Or I could just show the complete files at all references. Which one do you prefer?

Reply:
oh, i see! that's great, I think these anchors are fine